### PR TITLE
Adding Email Management

### PR DESCRIPTION
**Pull request by madison-boman**
Add the ability to manage email templates and send bulk emails.

### CODE REVIEW

File: src/GRA.Data/Repository/EmailTemplateRepository.cs [hunk 1]
```diff
@@ -11,5 +19,25 @@ public class EmailTemplateRepository
...
         {
         }
+
+        public async Task<DataWithCount<ICollection<EmailTemplate>>> PageAsync(BaseFilter filter)
+        {
+            var templates = DbSet.AsNoTracking();
+
+            var count = await templates.CountAsync();
+
+            var performerList = await templates
```
Comment (haraldnagel) [inline src/GRA.Data/Repository/EmailTemplateRepository.cs]: performerList?

File: src/GRA.Controllers/MissionControl/EmailManagementController.cs [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        {
+            var filter = new BaseFilter();
+            var templateList = await _emailManagementService.GetPaginatedEmailTemplateListAsync(filter);
+            var paginateModel = new PaginateViewModel
+            {
+                ItemCount = templateList.Count,
+                CurrentPage = page,
+                ItemsPerPage = filter.Take.Value
+            };
+            if (paginateModel.MaxPage > 0 && paginateModel.CurrentPage > paginateModel.MaxPage)
```
Comment (haraldnagel) [inline src/GRA.Controllers/MissionControl/EmailManagementController.cs]: I think there's a method in paginateModel called PastMaxPage or something that should be used here.

File: src/GRA.Domain.Service/EmailManagementService.cs [hunk 3]
```diff
@@ -13,18 +14,21 @@ public class EmailManagementService : BaseUserService<EmailManagementService>
...
             IUserContextProvider userContextProvider,
             IEmailSubscriptionAuditLogRepository emailSubscriptionAuditLogRepository,
+            IEmailTemplateRepository emailTemplateRepository,
             IUserRepository userRepository,
             IStringLocalizer<Resources.Shared> sharedLocalizer)
             : base(logger, dateTimeProvider, userContextProvider)
         {
             _emailSubscriptionAuditLogRepository = emailSubscriptionAuditLogRepository
                 ?? throw new ArgumentNullException(nameof(emailSubscriptionAuditLogRepository));
+            _emailTemplateRepository = emailTemplateRepository ?? throw new ArgumentNullException(nameof(emailTemplateRepository));
```
Comment (haraldnagel) [inline src/GRA.Domain.Service/EmailManagementService.cs]: Wrap lines at 100 characters

File: src/GRA.Domain.Service/EmailManagementService.cs [hunk 4]
```diff
@@ -104,5 +108,43 @@ public async Task TokenUnsubscribe(string token)
...
+        public async Task<DataWithCount<ICollection<EmailTemplate>>>
+            GetPaginatedEmailTemplateListAsync(BaseFilter filter)
+        {
+            if (HasPermission(Permission.ManageBulkEmails))
+            {
+                return await _emailTemplateRepository.PageAsync(filter);
+            }
+            else
+            {
+                _logger.LogError($"User {GetClaimId(ClaimType.UserId)} doesn't have permission to view the email template list.");
```
Comment (haraldnagel) [inline src/GRA.Domain.Service/EmailManagementService.cs]: Use structured logging, not interpolated strings

File: src/GRA.Domain.Service/EmailManagementService.cs [hunk 5]
```diff
@@ -104,5 +108,43 @@ public async Task TokenUnsubscribe(string token)
...
+                emailTemplate.EmailsSent = 0;
+                emailTemplate.FromAddress = emailTemplate.FromAddress.Trim();
+                emailTemplate.FromName = emailTemplate.FromName.Trim();
+                emailTemplate.SiteId = userId;
+                emailTemplate.Subject = emailTemplate.Subject.Trim();
+                return await _emailTemplateRepository.AddSaveAsync(userId, emailTemplate);
+            }
+            else
+            {
+                _logger.LogError($"User {GetClaimId(ClaimType.UserId)} doesn't have permission to create an email template.");
```
Comment (haraldnagel) [inline src/GRA.Domain.Service/EmailManagementService.cs]: Structured logging

File: src/GRA.Domain.Service/EmailManagementService.cs (hunk 4 above)
Comment (madison-boman) [reply src/GRA.Domain.Service/EmailManagementService.cs]: Fixed and noted

File: src/GRA.Domain.Service/EmailManagementService.cs (hunk 5 above)
Comment (madison-boman) [reply src/GRA.Domain.Service/EmailManagementService.cs]: Fixed

File: src/GRA.Domain.Service/EmailManagementService.cs (hunk 3 above)
Comment (madison-boman) [reply src/GRA.Domain.Service/EmailManagementService.cs]: fixed

File: src/GRA.Controllers/MissionControl/EmailManagementController.cs (hunk 2 above)
Comment (madison-boman) [reply src/GRA.Controllers/MissionControl/EmailManagementController.cs]: Got it

File: src/GRA.Data/Repository/EmailTemplateRepository.cs (hunk 1 above)
Comment (madison-boman) [reply src/GRA.Data/Repository/EmailTemplateRepository.cs]: Fixed

**Pull request merged**